packages/api/src/modules/providers/gispro/gispro.service.ts

```diff
@@ -0,0 +1,94 @@
+import GisproActionEntity from './entities/GisproActionEntity';
+import gisproRepository from './repositories/gispro.repository';
+import { isSiret, isAssociationName } from "../../../shared/Validators";
+import { DefaultObject, Siren, Siret } from '../../../@types';
+import GisproRequestAdapter from './adapters/GisproRequestAdapter';
+import DemandesSubventionsProvider from "../../demandes_subventions/interfaces/DemandesSubventionsProvider";
+import DemandeSubvention from '@api-subventions-asso/dto/search/DemandeSubventionDto';
+
+export const VALID_REQUEST_ERROR_CODE = {
+    INVALID_SIRET: 1,
+    INVALID_NAME: 2
+}
+
+export class GisproService implements DemandesSubventionsProvider {
+    public validEntity(entity: GisproActionEntity) {
+        if (!isSiret(entity.providerInformations.siret)) {
+            return { success: false, message: `INVALID SIRET FOR ${entity.providerInformations.siret}`, data: entity.providerInformations, code: VALID_REQUEST_ERROR_CODE.INVALID_SIRET };
+        }
+
+        if (!isAssociationName(entity.providerInformations.tier)) {
+            return { success: false, message: `INVALID NAME FOR ${entity.providerInformations.tier}`, data: entity.providerInformations, code: VALID_REQUEST_ERROR_CODE.INVALID_NAME };
+        }
+
+        return { success: true };
+    }
+
+    public async upsertMany(requests: GisproActionEntity[]) {
+        return await gisproRepository.upsertMany(requests);
+    }
+
+    public async insertMany(requests: GisproActionEntity[]) {
+        return await gisproRepository.insertMany(requests);
+    }
+
+    public async add(enitity: GisproActionEntity): Promise<{state: string, result: GisproActionEntity}> {
+        const existingFile = await gisproRepository.findByActionCode(enitity.providerInformations.codeAction);
+
+        if (existingFile) {
```

Comment:
Ce if qui update ou create est partagé par tous les services. On pourrait mutualiser ça je pense (en utilisant un this.repository qui définirait le repo du service et un this.existingFile utilisé dans la méthode add() du parent une fois définie par le service enfant)

Reply:
Je comprends le principe, mais j'avoue que la partie entre parenthèse avec les this.repo et this.existing sont un peu flou !

Reply:
Avec l'héritage de prototype tu peux avoir un parent avec un this.repository = null et l'enfant fait this.repository = OsirisRepository. Après le parent a un méthode add() qui fait le traitement `if (existingFile) { ... } else { ... }`. En gros ça donnerait: 
` // GisproService extends ProviderService
this.repository = GisproRepository;
public async add(entity) { 
  this.existingFile = await gisproRepository.findByActionCode(enitity.providerInformations.codeAction);
  super()
}

// ProviderService
this.repository = null;
public async add (entity) { 
 if (this.existingFile) { 
  this.repository.update(entity)
 } else { 
  this.repository.add(entity)
 }
}

Reply:
Pour le add pourquoi pas, maisp pour le existing, c'est plus chaud parce que tous les provider non pas le même identifiant unique ! Mais peut être qu'on pourrait mettre un unique id dans chaque entité et ducoup ce basé dessus pour le existing